python bindings: use meson-python

arthurzam opened the pull request:
@amaanq Hi
After various fights against setuptools and cffi (including new generator you pushed at https://github.com/tree-sitter/tree-sitter/pull/2438/commits/2aa4838fbeb453255e007a1cbbfc0faf407c211e) it was all failing for me.

By some recommendations from @eli-schwartz (meson author, and a great help at Gentoo), I've taken my luck with cffi. And look at this PR, it works well.

The best test you can do to have it as isolated as possible:

```sh
# create new virtual env & activate it
python3 -m venv /tmp/venv
source /tmp/venv/bin/activate
# install from source (builds wheel and installs it)
cd tree-sitter-capnp
pip install ./bindings/python
# test it: "from tree_sitter_capnp import language"
python
```

## Caveats

1. This isn't using the best abi variant for our case (abi3). meson live supports it, and when they release it it would be very easy to add (just add one line to `meson.build`).

   What this means for us? For now you would need to generate a wheel file for each python target, which can be done quite easily in CI (I could help with it). When new meson is released, we update the generator and we would be able to build one wheel for all python versions (starting from minimal 3.x we can decide) - so only matrix is for arches and glibc/musl.

2. My knowledge of rust is non-existent (I can somehow read it, since it is code, but not write) so I can't send PR against the generator. What I can do is create this PR, which brings you the expected output and you change the generator.

arthurzam committed:
python bindings: use meson-python

amaanq commented:
Hi, nice to see you're keeping up w/ me on this! - this branch was definitely outdated and didn't regenerate with my latest commit to the bindings PR

I just updated *my* branch and it's slightly different - I cast the cffi function pointer to an int, I recently realized the Language class in py-tree-sitter is an int, returning a c_void_p with cdll gives you an int in Python. So, some changes to py-tree-sitter *would* be made to accomadate for this, but, once I do make those, we can just do this with this branch:

```py
from tree_sitter import Language, Parser
import tree_sitter_capnp

CAPNP = Language(tree_sitter_capnp.language(), 'capnp')

parser = Parser()
parser.set_language(CAPNP)

tree = parser.parse(b"""
@0x0000000000000000;
struct Foo {
    bar @0 :Int64;
    baz @1 :Int64;
}
""")
print(tree.root_node.sexp())
```

Which is *exactly* how I've wanted it to work, output:

```sexp
(message (unique_id_statement (unique_id)) (struct (type_identifier) (field (field_identifier) (field_version) (field_type (primitive_type))) (field (field_identifier) (field_version) (field_type (primitive_type)))))
```

arthurzam commented:
> Hi, nice to see you're keeping up w/ me on this! - this branch was definitely outdated and didn't regenerate with my latest commit to the bindings PR

Well, of course I'll follow it up. While I don't have too much time to work on it, if I said I'll do something I'll do it or bring notice that I gave up 😄 Also I really want this to work and be used for tree-sitter-bash, since it would make my life with pkgcheck tool much easier (currently we perform a hack to supply the .so file in the wheel, but this is an ugly hack).

> I just updated my branch and it's slightly different - I cast the cffi function pointer to an int, I recently realized the Language class in py-tree-sitter is an int, returning a c_void_p with cdll gives you an int in Python. So, some changes to py-tree-sitter would be made to accommodate for this, but, once I do make those, we can just do this with this branch:

No issues, I've updated my branch to apply your changes and mine above. Your variant still fails for me, while the current one (mine) works at all known scenarios for me. It is hard for me to see if your ideas about language being an int will work well or such, but I'll leave it to you 😄 

> Which is exactly how I've wanted it to work, output:

Looks great!

>  having auto-built wheels from the CI would be great too!

I agree with your plan, just for starters we should bring the manual commands to build the wheel, and then work on automation. I'll gladly help with that (and how to set it up securely in github).

eli-schwartz commented:
> By some recommendations from @eli-schwartz (meson author, and a great help at Gentoo), I've taken my luck with cffi. And look at this PR, it works well.

:o Well, happy to hear that it is working out for you, then.

eli-schwartz commented on bindings/python/meson.build:
```diff
@@ -0,0 +1,26 @@
+  'tree-sitter-capnp',
+  'c',
+  license: 'MIT'
+)
```
I would add also:

```meson
meson_version: '>=0.64.0',
```
since you are using functionality that depends on this. While pip install will typically always install the latest version, this lets meson produce a better error message e.g. on LTS distros with an older meson available.

If you set this to `meson_version: '>=0.37.0'` (chosen for no particular reason other than that's when we started tracking features) you get this warning:

```
meson.build:8: WARNING: Project targets '>=0.37.0' but uses feature introduced in '0.46.0': module python.
meson.build:8: WARNING: Project targets '>=0.37.0' but uses feature introduced in '0.51.0': modules arg in python.find_installation.
meson.build:8: WARNING: Project targets '>=0.37.0' but uses feature introduced in '0.64.0': pure arg in python.find_installation.

[...]

meson.build:19: WARNING: Project targets '>=0.37.0' but uses feature introduced in '0.49.0': / with string arguments.
meson.build:19: WARNING: Project targets '>=0.37.0' but uses feature introduced in '0.49.0': / with string arguments.
meson.build:19: WARNING: Project targets '>=0.37.0' but uses feature introduced in '0.49.0': / with string arguments.
meson.build:16: WARNING: Project targets '>=0.37.0' but uses feature introduced in '0.48.0': gnu_symbol_visibility arg in build target.
meson.build:27: WARNING: Project targets '>=0.37.0' but uses feature introduced in '0.49.0': / with string arguments.
```

and a final summary:

```
WARNING: Project specifies a minimum meson_version '>=0.37.0' but uses features which were added in newer versions:
 * 0.46.0: {'module python'}
 * 0.48.0: {'gnu_symbol_visibility arg in build target'}
 * 0.49.0: {'/ with string arguments'}
 * 0.51.0: {'modules arg in python.find_installation'}
 * 0.64.0: {'pure arg in python.find_installation'}
```
This makes it easy to quickly check which version of the meson.build API you need.

eli-schwartz commented on bindings/python/meson.build:
```diff
@@ -0,0 +1,26 @@
+    files('..' / '..' / 'src' / 'parser.c'),
+    # if you have a scanner, uncomment next line
+    # files('..' / '..' / 'src' / 'scanner.c'),
+    dependencies: [python_dep],
```
Since you require 0.64.0 anyway, you can drop this explicit dependency on python.dependency(), it is automatically added for you if you don't provide one since 0.63.0.

eli-schwartz commented on bindings/python/meson.build:
```diff
@@ -0,0 +1,26 @@
+    subdir: 'tree_sitter_capnp',
+)
+
+python.install_sources(files('tree_sitter_capnp' / '__init__.py'), subdir: 'tree_sitter_capnp')
```
The main value of files() is for obtaining cross-directory references saved as reusable variables, that remember the directory in which they were defined. Passing strings to install_sources() is also valid, and internally converts immediately to a `mesonbuild.mesonlib.File` object.

(The second value of it is for cases where a function accepts command strings, not "list of source files", and you need to make sure meson understands it is a file, not "just" an argument to a command.)

It is slightly shorter, and identical in effect, so why not do a bit less typing. :D

Another thing available since 0.64.0:

```meson
python.install_sources('tree_sitter_capnp/__init__.py'), preserve_path: true)
```

will install the entire structure as defined, so if you installed the "tree_sitter_capnp/__init__.py" file it will install the python module "tree_sitter_capnp" subdir's file "__init__.py".

eli-schwartz commented on bindings/python/meson.build:
```diff
@@ -0,0 +1,26 @@
+ts_pyext = python.extension_module(
+    '_tree_sitter_capnp',
+    ts_cffi_srcs,
+    files('..' / '..' / 'src' / 'parser.c'),
```
(files objects aren't strictly necessary here, again, since it's a parameter to a function that explicitly takes files as either string or files())

Note that meson doesn't require you to use the `/` path-joining operator for statically coded paths like this, internally it just operates on unix-style forward-slash paths. The path-joining operator is mainly *useful* for handling cases where you add together paths that you don't control, which may be absolute and may or may not have trailing "/", which causes string `+` concatenation to potentially cause bad things.

Here you just have one path segment: `'../../src/parser.c'`.

eli-schwartz commented:
> 1. What this means for us? For now you would need to generate a wheel file for each python target, which can be done quite easily in CI (I could help with it). When new meson is released, we update the generator and we would be able to build one wheel for all python versions (starting from minimal 3.x we can decide) - so only matrix is for arches and glibc/musl.

If you really wanted to here, you could do this:

```meson
if meson.version().version_compare('>=1.3.0')
    addedkwargs = {'limited_api': '3.7'}
else
    addedkwargs = {}
endif
```
ts_pyext = python.extension_module(
    '_tree_sitter_capnp',
    ts_cffi_srcs,
    files('..' / '..' / 'src' / 'parser.c'),
    # if you have a scanner, uncomment next line
    # files('..' / '..' / 'src' / 'scanner.c'),
    dependencies: [python_dep],
    install: true,
    subdir: 'tree_sitter_capnp',
    kwargs: addedkwargs,     # this is magic
)
```

Meson is a bit dumb and can't tell that the dictionary is only passing the limited_api kwarg on new enough versions of meson -- so it *does* emit a warning telling you that you used a 1.3.0 feature when you support 0.64.0. But it also means you automatically get support as soon as meson releases the feature.

Up to you which one you care more about. :)

amaanq commented:
> No issues, I've updated my branch to apply your changes and mine above. Your variant still fails for me, while the current one (mine) works at all known scenarios for me. It is hard for me to see if your ideas about language being an int will work well or such, but I'll leave it to you 😄

It's likely because your Language() constructor doesn't take in the overloaded version I've added, I can make a py-tree-sitter fork with my local changes for you to try it out as well if you'd like :smile:

amaanq commented:
Here's the fork, I just pushed what I had and iirc it's pretty much done https://github.com/tree-sitter/py-tree-sitter/tree/prepare-bindings

Sorry for not responding yesterday, I lost power due to a storm :grimacing:

eli-schwartz commented:
> But tbh, I'm not all that sure about using meson-python, does that actually use meson under the hood? To me cffi/cdll is simpler and preferable if possible

It replaces setup.py with meson, the use of cffi/cdll is unaffected AFAICT.
